cpp/include/cudf/scalar/scalar_device_view.cuh

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/string_view.cuh>
+#include <cudf/types.hpp>
+
+namespace cudf {
+
+namespace detail {
+
+/**---------------------------------------------------------------------------*
+ * @brief An immutable, non-owning view of scalar from device that is trivially 
+ * copyable and usable in CUDA device code.
+ *---------------------------------------------------------------------------**/
+class alignas(16) scalar_device_view_base {
+ public:
+  scalar_device_view_base() = default;
+  ~scalar_device_view_base() = default;
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the element type
+   *---------------------------------------------------------------------------**/
+  __host__ __device__ data_type type() const noexcept { return _type; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns whether the scalar holds a valid value (i.e., not null).
+   *
+   * @return true The element is valid
+   * @return false The element is null
+   *---------------------------------------------------------------------------**/
+  __device__ bool is_valid() const noexcept {
+    return *_is_valid;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns whether the value is null.
+   *
+   * @return true The element is null
+   * @return false The element is valid
+   *---------------------------------------------------------------------------**/
+  __device__ bool is_null() const noexcept {
```

Comment:
I don't see the need for both is_null and is_valid. Also, at the column level we have APIs `is_null` and `is_not_null`, so this is inconsistent. Why does `scalar` not have both, while `scalar_device_view` has both?

Reply:
I already suggested renaming the `is_not_null` column API to `is_valid`.

`column_device_view` has `is_null` and `is_valid`. I think `is_valid()` reads more naturally than `not is_null()`.

Reply:
`is_null` and `is_valid` exist for consistency with `cudf::column`. Similarly for `set_valid` and `set_null`. I can remove the extra setter but for consistency, I'd have to remove the extra getter. Is that ok? So I'm thinking of just keeping `set_valid(bool)` and `is_valid()`